### PR TITLE
Issue 484

### PR DESCRIPTION
**Pull request by mickaelistria**

### CODE REVIEW

**Comment by laeubi** [comment]
@mickaelistria actually I really like to get rid of the BlackboardRepository and there is a TODO in the code about it, do have found out why the Director needs this? Is this some legacy code? Or should we add some more documentation to the BlackboardRepository stuff do document whats going on?

**Comment by laeubi** [comment]
Okay maybe I have misunderstood it does the "get rid of" todo means we should get rid of blackboard key(=not use it at all), or was it meant to say its implementation was/is incomplete? I really curious about that as I have stumbled too often about this TODO....

File: tycho-bundles/org.eclipse.tycho.p2.resolver.impl.test/src/test/java/org/eclipse/tycho/p2/target/TestResolverFactory.java [hunk 1]
```diff
@@ -47,6 +59,39 @@ public TestResolverFactory(MavenLogger logger) {
...
         targetDefinitionResolverService.setMavenContext(mavenContext);
+        targetDefinitionResolverService.setMavenDependenciesResolver(new MavenDependenciesResolver() {
+
+            @Override
+            public Collection<?> resolve(String groupId, String artifactId, String version, String packaging,
+                    String classifier, String dependencyScope, int depth,
+                    Collection<MavenArtifactRepositoryReference> additionalRepositories, Object session)
+                    throws DependencyResolutionException {
+                // This is supposed to mimic Maven repo returning an artifact
+                String relativePath = groupId.replace('.', File.separatorChar) + File.separator + artifactId
```
Comment (laeubi) [inline tycho-bundles/org.eclipse.tycho.p2.resolver.impl.test/src/test/java/org/eclipse/tycho/p2/target/TestResolverFactory.java:70]: Can we use GAV(....) here as it already has a method to construct a maven repo path?

File: tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetPlatformFactoryImpl.java [hunk 2]
```diff
@@ -186,18 +182,26 @@ public P2TargetPlatform createTargetPlatform(TargetPlatformConfigurationStub tpC
...
                 filter, //
                 localMetadataRepository, //
-                createExternalArtifactProvider(completeRepositories, targetFileContent, pomDependencyArtifactRepo,
-                        tpConfiguration.getIncludePackedArtifacts()), //
+                externalArtifactFileProvider, //
                 localArtifactRepository, //
                 includeLocalMavenRepo, //
                 logger);
 
+        RepositoryBlackboardKey blackboardKey = RepositoryBlackboardKey
```
Comment (laeubi) [inline tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetPlatformFactoryImpl.java:199]: Can we add a little description what this blackboardkey is for and what it does (beside that some test fails if it is not there...). Is it some legacy code we might remove in the future?

File: tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetPlatformFactoryImpl.java [hunk 3]
```diff
@@ -307,28 +311,60 @@ private IMetadataRepository loadMetadataRepository(MavenRepositoryLocation locat
         }
     }
 
+    private static final class SortedRepositories {
```
Comment (laeubi) [inline tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetPlatformFactoryImpl.java:314]: can you add a little comment by what mean it is sorted? I assume its local first, then remote or something?

**Review by laeubi — COMMENTED** [review]
Looks good and if it resolves the issue of mirroring local items that's a great improvement.

**Comment by mickaelistria** [comment]
> actually I really like to get rid of the BlackboardRepository and there is a TODO in the code about it

Me too, one of my favorite things.

> do have found out why the Director needs this? Is this some legacy code? Or should we add some more documentation to the BlackboardRepository stuff do document whats going on?

The blackboard repository allows to keep the project TargetPlatform artifact repositories in memory and maps it with a URL of form `file:/resolution-context-artifacts@%2Fpath%2Fto%2Fproject`, that the director mojo can then use to retrieve the dependencies. This is partly documented on `RepositoryBlackboardKey#forResolutionContextArtifacts`.

By "get rid of the todo" I mean fix the implementation as suggested (ie resolution context now contains the whole Target Platform artifacts and not only the pom-dependencies.

**Comment by laeubi** [comment]
But why can't we pass the memory reference some other way? Is this a restriction of the org.eclipse.equinox.internal.p2.director.app.DirectorApplication.initializeRepositories() being private? Interestingly metadataManager is already protected so one should be able to set this directly...

Anyways we should merge this and take a look at the blackboard repository later, can you create an issue that describes why we currently need this (e.g. because of DirectorApplication restriction or something else...)

**Comment by mickaelistria** [comment]
> But why can't we pass the memory reference some other way?

DirectorApplication wants a URL, so this BlackboardRepositoryKey builds an URL which resolves to an ArtifactRepositoryBlackboard containing the target platform.

File: tycho-bundles/org.eclipse.tycho.p2.resolver.impl.test/src/test/java/org/eclipse/tycho/p2/target/TestResolverFactory.java (hunk 1 above)
Comment (mickaelistria) [reply tycho-bundles/org.eclipse.tycho.p2.resolver.impl.test/src/test/java/org/eclipse/tycho/p2/target/TestResolverFactory.java:70]: The only GAV accessible here doesn't have method to build a path as far as I could see. Or did I miss something>

**Pull request merged**